### ISSUE TITLE
"*" is not equal to any()

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/semver#161]
I would expect the following program to print `* * true`. Instead it prints `* * false`.

```rust
extern crate semver;
use semver::VersionReq;

fn main() {
    let any = VersionReq::any();
    let star = VersionReq::parse("*").unwrap();
    println!("{} {} {}", any, star, any == star);
}
```

**Comment by KodrAus** [comment]
Looks like we might need to implement equality traits manually by calling `matches` instead of deriving them?

**Comment by connorkuehl** [comment]
I don't think there's anything wrong with the equality traits. It looks like `VersionReq` derives it all the way down.

I added:

```Rust
println!("{:?}\n{:?}", any, star);
```

to @dtolnay's test program.

The output is interesting:

```Rust
* * false
VersionReq { predicates: [Predicate { op: Wildcard(Major), major: 0, minor: None, patch: None, pre: [] }] }
VersionReq { predicates: [] }
```

Definitely not the same 😨, but it should be.

I've spent some time trying to debug this but `rust-lldb` is going to give me an actual heart palpation. I'm having a horrible time setting breakpoints. That's a rant for some place else, though.

I wonder if it could be something to do with the crate boundary between `semver` and `semver_parser`? `From::from()` for `semver::VersionReq` and `semver_parser::VersionReq`? That code looks fairly standard, though.

`VersionReq::parse("*")` should lead to a call to `semver_parser::range::parse(input)` which should return a `VersionReq` with an empty predicates vec (one that must be translated from a `sever_parser::VersionReq` into a `semver::VersionReq`. I'm inclined to believe we have a stowaway predicate somewhere around here. Unfortunately without my debugger actually being able to break on any of these I can't find exactly where that weird predicate is being inserted. 

tl;dr -- `VersionReq::parse("*")` is producing an `VersionReq` with a phantom predicate; the "any" VersionReqs being considered here are not consistent with each other.

**Comment by agalakhov** [comment]
It's even worse: the star does not work as expected.

```rust
let req = VersionReq::parse("*").unwrap();
let vers = Version::parse("0.1.0-beta");
assert!(req.matches(&vers)); // ??? !!!
```

**Comment by udoprog** [comment]
Hey,

This is fixed in #151, and is also partly the motivation for why #159 was introduced.

**Comment by matklad** [comment]
So, as pointed out by @agalakhov , `*` and `any` are indeed different, and to me it looks like this might be intentional?

The `*` is used for, at least in Cargo, "give me the latest version of a crate", where "latest" does not include prereleases. `any` is used internally to mean "I really don't care about version requirements at all, and prereleases are fine with me".

**Comment by dtolnay** [comment]
This is fixed in 1.0.0. "Any" is no longer a thing, because there is no SemVer syntax that behaves that way. "Star" is a different thing and is exposed as https://docs.rs/semver/1.0.0-rc.1/semver/struct.VersionReq.html#associatedconstant.STAR. If someone needs the semantics of sometimes always matching everything, then they should use `Option<VersionReq>`, and `.map_or(true, |req| req.matches(version))`.
